README.md
# node-huaban
花瓣(http://huaban.com) 画板下载器

[![npm version](https://img.shields.io/npm/v/node-huaban.svg?style=flat-square)](#)
[![node version](https://img.shields.io/node/v/node-huaban.svg?style=flat-square)](#)
[![license](https://img.shields.io/npm/l/node-huaban.svg?style=flat-square)](#)


## 安装

```
$ npm i node-huaban -g
```

## 使用

```
$ huaban
  huaban board downloader v0.0.1

    Usage:
      huaban [options] <board_url>

    Options:
      -h,--help         输出此帮助信息
      -c,--concurrency  同时最大下载数量,默认10
      -t, --timeout     超时, 单位分, 默认1分钟 = 60 * 1000毫秒
      --max-times       重试次数, 默认5

    Example:
      huaban http://huaban.com/boards/17324249/ -c 10
```

## API

```js
const HuabanBoard = require('node-huaban');
```

### constructor

```js
const board = new HuabanBoard(url) // url 是画板地址
```

### init

```js
board.init() // return a Promise, when all pins & title are ready

// normally, use with async/await
await board.init()
```

运行 `init` 之后, board会包含以下字段:
- board.title board的名称
- board.pins 图片, 包含地址src & 文件类型 ext

### downloadBoard
下载所有图片

```js
await board.downloadBoard(concurrency, timeout, maxTimes)
```

## Changelog

[CHANGELOG.md](CHANGELOG.md)

## License

the MIT License http://magicdawn.mit-license.org

CHANGELOG.md
# Changelog

## 2018-04-07 v1.4.1

- 合并 PR #3, 解决画板名称特殊字符问题

## 2018-02-22 v1.4.0

- 合并 PR #2, 解决字符串中包含 `{` 或 `}` 的问题
- 使用 prettier 代替 js-beautify
- 使用 fs-extra@5 自带 promise 支持, 移除 promise.ify
- 使用 mime@2
- 使用 async/await, 移除 co, 要求 node >= 7.6.0

## 2016-11-06 v1.3.0
- 重构代码, 去掉 fs-extra-promise, 使用 promise.ify & fs-extra
- 重构代码, 去掉 superagent, 使用 request & request-promise

## 2016-08-30 v1.2.0
- 升级依赖
- 重构代码
- 去掉 `needle-kit`

## 2016-03-09 v1.1.0
- 移除 bluebird 依赖, 使用ES6 原生 Promise

## Unknown date v1.0.0
- 初始版本

lib/util.js
const path = require('path')
const fs = require('fs')
const fse = require('fs-extra')
const pretry = require('promise.retry')
const r = require('request')
const rp = require('request-promise').defaults({
  simple: false,
})

/**
 * get html for url
 */

exports.getHtml = function(url, query) {
  return rp.get(url, {
    qs: query,
  })
}

/**
 * match right
 */

exports.getRight = function(input, fi) {
  const pairs = {
    '(': ')',
    '{': '}',
    '[': ']',
  }

  const left = input[fi]
  const right = pairs[left]

  if (!right) {
    return -1
  }

  let count = 1 // input[fi] = left

  for (let i = fi + 1, len = input.length; i < len; i++) {
    const cur = input[i]
    if (cur === right) {
      count--
      if (count === 0) {
        return i
      }
    } else if (cur === left) {
      count++
    }
  }

  return -1 // not found
}

/**
 * 下载一个文件
 */

exports.download = (src, dest, onCancel) =>
  new Promise((resolve, reject) => {
    // ensure
    dest = path.resolve(dest)
    fse.ensureDirSync(path.dirname(dest))
    const s = fs.createWriteStream(dest)

    // construct
    const req = r.get(src).on('error', reject)

    // pipe
    req
      .pipe(s)
      .on('error', reject)
      .on('finish', function() {
        this.close(() => {
          resolve()
        })
      })

    onCancel &&
      onCancel(() => {
        req.abort()
        s.close()
      })
  })

/**
 * tryDownload with timeout & times
 */

exports.tryDownload = async function(src, dest, timeout, times) {
  const fn = pretry(exports.download, {
    times,
    timeout,
    onerror: (err, index) => {
      // console.error('obj');
    },
  })

  await fn(src, dest)
}


lib/index.js
const path = require('path')
const fse = require('fs-extra')
const sanitizeFilename = require('sanitize-filename')
const _ = require('lodash')
const mime = require('mime')
const debug = require('debug')('huaban:index')
const logSymbols = require('log-symbols')
const pmap = require('promise.map')
const util = require('./util')

/**
 * image host
 */

const imgHosts = {
  hbimg: 'img.hb.aicdn.com',
  hbfile: 'hbfile.b0.upaiyun.com',
  hbimg2: 'hbimg2.b0.upaiyun.com',
}

/**
 * get board data
 */

const getBoard = function(html) {
  // debug('html in getBoard: %s', html);

  /* eslint quotes: 0 */
  const search = `app.page["board"] = {`
  const start = html.indexOf(search) + search.length - 1 // html[start] = '{'
  //const end = util.getRight(html, start) // html[end] = '}';
  const end = html.indexOf('};\n', start)
  const js = html.slice(start, end + 1)
  // debug('js in getBoard: %s', js);

  return JSON.parse(js)
}

/**
 * get pins on a page
 */

const getPagePins = function(html) {
  const board = getBoard(html)
  return board.pins
}

/**
 * get pins from url、total amount
 */

const getPins = async (url, total) => {
  let pins = []
  if (url.slice(-1) !== '/') url += '/'

  let i = 1
  total = Math.ceil(total / 100) // e.g 150 need 2 requests
  while (i <= total) {
    const query = {
      limit: 100,
    }

    if (pins.length > 0) {
      // not the first time
      query.max = pins[pins.length - 1].pin_id
    }

    const html = await util.getHtml(url, query)
    const _pins = getPagePins(html)
    pins = pins.concat(_pins)

    i++
  }

  return pins.map((pin, index) => {
    return {
      src: `http://${imgHosts[pin.file.bucket]}/${pin.file.key}`,
      ext: mime.getExtension(pin.file.type) || 'jpg',
    }
  })
}

module.exports = class HuabanBoard {
  constructor(url) {
    this.url = url
    this.title = undefined
    this.pins = undefined
  }

  async init() {
    const html = await util.getHtml(this.url)
    const board = getBoard(html)

    // title
    this.title = board.title
    this._raw = board

    this.id = board.board_id

    const pins = await getPins(this.url, board['pin_count'])
    return (this.pins = pins)
  }

  async downloadBoard(concurrency, timeout, maxTimes) {
    // dir
    const title = this.title
    const id = this.id
    const folderName = sanitizeFilename(`${id}-${title}`)
    const dir = path.resolve(folderName)
    fse.ensureDirSync(dir)

    // pad length
    const numLength = String(this.pins.length).length
    const length = this.pins.length

    return await pmap(
      this.pins,
      function(pin, index) {
        const num = _.padStart(String(index + 1), numLength, '0') // 001
        const dest = path.join(dir, `${num}.${pin.ext}`)

        return util
          .tryDownload(pin.src, dest, timeout, maxTimes)
          .then(function() {
            console.log(`${logSymbols.success} %s/%s 下载成功`, num, length)
          })
          .catch(function(e) {
            console.error(`${logSymbols.error} %s/%s 下载失败`, num, length)
            console.error(e.stack || e)
          })
      },
      concurrency
    )
  }
}


huaban.js
#!/usr/bin/env iojs

/**
 * module dependencies
 */
// 3rd
global.Promise = require('bluebird');
var co = require('co');
var request = require('superagent');
require('superagent-bluebird-promise');
var _ = require('lodash');
var mime = require('mime');
var minimist = require('minimist');
// core lib
var fmt = require('util').format;
var fs = require('fs');
var path = require('path');

/**
 * example url
 */
var exampleUrl = 'http://huaban.com/boards/17324249/'


/**
 * image host
 */
var imgHosts = {
  "hbimg": "img.hb.aicdn.com",
  "hbfile": "hbfile.b0.upaiyun.com",
  "hbimg2": "hbimg2.b0.upaiyun.com"
};


/**
 * get html for url
 */
var getHtmlAsync = function(url, query) {
  /** construct request */
  var req = request
    .get(url)

  if (query) {
    req.query(query);
  }

  return req
    .promise()
    .then(function(res) {
      return res.text;
    })
};


/**
 * match right
 */
var getRight = function(input, fi) {
  var pairs = {
    '(': ')',
    '{': '}',
    '[': ']'
  };

  var left = input[fi];
  var right = pairs[left];

  if (!right) {
    return -1;
  }

  var count = 1; // input[fi] = left

  for (var i = fi + 1, len = input.length; i < len; i++) {
    var cur = input[i];

    if (cur === right) {
      count--;
      if (count === 0) {
        return i;
      }
    } else if (cur === left) {
      count++;
    }
  }

  return -1; // not found
};


/**
 * get board data
 */
var getBoard = function(html) {
  var search = 'app.page["board"] = {';
  var start = html.indexOf(search) + search.length - 1; // html[start] = '{'
  var end = getRight(html, start); // html[end] = '}';
  var js = html.slice(start, end + 1);

  return JSON.parse(js);
};


/**
 * get pins on a page
 */
var getPagePins = function(html) {
  var board = getBoard(html);
  return board.pins;
}


/**
 * get pins from url、total amount
 */
var getPinsAsync = co.wrap(function*(url, total) {
  var pins = [];
  if (url.slice(-1) !== '/') url += '/';

  var i = 1;
  total = Math.ceil(total / 100); // e.g 150 need 2 requests
  while (i <= total) {
    var query = {
      limit: 100
    };

    if (pins.length > 0) { // not the first time
      query.max = pins[pins.length - 1].pin_id;
    }

    var html =
      yield getHtmlAsync(url, query);
    var _pins = getPagePins(html);
    pins = pins.concat(_pins);

    i++;
  }

  // image download needs src&dest
  var fileNameLength = String(pins.length).length;
  pins = pins.map(function(pin, index) {
    var _pin = {
      src: fmt('http://%s/%s', imgHosts[pin.file.bucket], pin.file.key),

      index: index, // 0
      num: _.padLeft(String(index + 1), fileNameLength, '0'), // 001

      ext: mime.extension(pin.file.type) || 'jpg'
    }

    _pin.dest = _pin.num + '.' + _pin.ext;
    return _pin;
  });

  return pins;
});

var downloadAsync = function(src, dest) {
  return new Promise(function(resolve, reject) {
    request
      .get(src)
      .on('error', reject)
      .pipe(fs.createWriteStream(dest))
      .on('error', reject)
      .on('finish', function() {
        this.close(function() { // close filestream
          resolve();
        })
      })
  });
};

/**
 * from http://huaban.com/board/xxx download all images(pins)
 */
var downloadBoardAsync = co.wrap(function*(url, concurrency) {
  var html =
    yield getHtmlAsync(url);
  var board = getBoard(html);

  /** output board data */
  console.log('board title : %s', board.title);
  console.log('board id : %d', board.board_id);
  console.log('board pins count : %d', board.pin_count);

  // all pins
  var pins =
    yield getPinsAsync(url, board.pin_count);

  // take care of dir
  var dir = path.resolve(board.title);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir);

  // start download
  console.log('Start download ...');
  yield Promise.map(pins, function(pin, index, length) {
    return downloadAsync(pin.src, path.join(dir, pin.dest))
      .then(function() {
        console.log("√ %s/%s 下载成功", pin.num, length);
      })
      .catch(function() {
        console.error("× %s/%s 下载失败", pin.num, length);
      })
  }, {
    concurrency: concurrency
  });

});

var argv = minimist(process.argv.slice(2), {
  boolean: ['help'],
  alias: {
    c: ['concurrency'],
    h: ['help']
  },
  defaults: {
    concurrency: 10
  }
});

if (process.argv.length === 2 || argv.help) {
  console.log(`

  huaban board downloader

    Usage:
      iojs huaban.js <board-url> [options]

    Options:
      -c,--concurrency    ->    同时最大下载数量,默认10
      -h,--help           ->    输出此帮助信息

    Example:
      iojs huaban.js ${ exampleUrl } -c 10
  `);

  process.exit();
} else {
  co(function*() {
    var url = argv._[0];
    var concurrency = argv.concurrency;

    var start = new Date;
    yield downloadBoardAsync(url, concurrency);
    var end = new Date;
    console.log('耗时 %s 秒', (end - start) / 1000);
  }).catch(function(e) {
    console.error(e);
  })
}